evm/p2p/auth.py

```diff
@@ -0,0 +1,284 @@
+        # egress-mac = sha3.update(mac-secret ^ recipient-nonce || auth-sent-init)
+        mac1 = keccak_updateable(sxor(mac_secret, responder_nonce) + auth_init)
+        # ingress-mac = sha3.update(mac-secret ^ initiator-nonce || auth-recvd-ack)
+        mac2 = keccak_updateable(sxor(mac_secret, initiator_nonce) + auth_ack)
+
+        if self._is_initiator:
+            egress_mac, ingress_mac = mac1, mac2
+        else:
+            egress_mac, ingress_mac = mac2, mac1
+
+        return aes_secret, mac_secret, egress_mac, ingress_mac
+
+
+class HandshakeInitiator(HandshakeBase):
+    _is_initiator = True
+
+    def encrypt_auth_message(self, auth_message):
+        auth_init = ecies.encrypt(auth_message, self.remote.pubkey)
+        return auth_init
+
+    def create_auth_message(self, nonce):
+        ecdh_shared_secret = ecies.ecdh_agree(self.privkey, self.remote.pubkey)
+        secret_xor_nonce = sxor(ecdh_shared_secret, nonce)
+
+        # S(ephemeral-privk, ecdh-shared-secret ^ nonce)
+        # FIXME: Should use ecdsa_sign(), but can't because it hashes the msg before signing.
+        S = encode_signature(
+            *get_ecc_backend().ecdsa_raw_sign(
+                secret_xor_nonce,
+                self.ephemeral_privkey)
+        )
+
+        # S || H(ephemeral-pubk) || pubk || nonce || 0x0
+        flag = 0x0
+        auth_message = S + keccak(self.ephemeral_pubkey) + self.pubkey + \
+            nonce + ascii_chr(flag)
+        return auth_message
+
+    def decode_auth_ack_message(self, ciphertext):
+        try:
+            eph_pubkey, nonce, version = decode_ack_plain(ciphertext, self.privkey)
+        except AuthenticationError:
+            eph_pubkey, nonce, version = decode_ack_eip8(ciphertext, self.privkey)
+            self.got_eip8_auth = True
+        auth_ack = ciphertext
+        return eph_pubkey, nonce, auth_ack
+
+
+class HandshakeResponder(HandshakeBase):
+    _is_initiator = False
+
+    def create_auth_ack_message(self, nonce):
+        if self.got_eip8_auth:
+            data = rlp.encode(
+                (self.ephemeral_pubkey, nonce, supported_rlpx_version),
+                sedes=eip8_ack_sedes)
+            # Pad with random amount of data. The amount needs to be at least 100 bytes to make
+            # the message distinguishable from pre-EIP-8 handshakes.
+            msg = data + os.urandom(random.randint(100, 250))
+        else:
+            # Unused, according to EIP-8, but must be included nevertheless.
+            token_flag = b'\x00'
+            msg = self.ephemeral_pubkey + nonce + token_flag
+        return msg
+
+    def encrypt_auth_ack_message(self, ack_message):
+        if self.got_eip8_auth:
+            # The EIP-8 version has an authenticated length prefix.
+            prefix = struct.pack('>H', len(ack_message) + ecies.encrypt_overhead_length)
+            auth_ack = ecies.encrypt(ack_message, self.remote.pubkey, shared_mac_data=prefix)
```

Comment:
Maybe rename to `suffix` to avoid the overwriting of this value on the next line.